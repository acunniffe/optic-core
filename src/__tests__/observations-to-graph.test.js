//@HINT: to visualize stuff, do something like require('fs').writeFileSync('output.graphviz.txt', observationsToGraph.graph.toGraphViz());

const { GraphQueries } = require('../graph/graph-queries');

const { ObservationsToGraphBuilder } = require('../observations-to-graph');
const { pathObservations, singleRequestBody } = require('./optic-observations.fixture.js');
const {toMatchSnapshot} = require('jest-snapshot');
expect.extend({toMatchSnapshot});

describe('ObservationsToGraph', function() {
  describe('simple API', function() {
    const observationsToGraph = ObservationsToGraphBuilder.fromEmptyGraph()
    observationsToGraph.interpretObservations(pathObservations);

    it('should understand the request body format', function() {
      observationsToGraph.interpretObservations(singleRequestBody({
        'emails': [
          'test@optic.com',
          'test2@optic.com',
        ],
      }));
      const graphQueries = new GraphQueries(observationsToGraph.graph);
      const [schemaRootNode] = graphQueries.listNodesByType('schemaRoot');
      expect(schemaRootNode.toJsonSchema()).toMatchSnapshot()
    });

    it('should understand the request body format with different observations', function() {
      observationsToGraph.interpretObservations(singleRequestBody({
        'emails': [],
      }));
      observationsToGraph.interpretObservations(singleRequestBody([]));
      observationsToGraph.interpretObservations(singleRequestBody('abc'));
      observationsToGraph.interpretObservations(singleRequestBody({
        'emails': [
          'test@optic.com',
          'test2@optic.com',
        ],
      }));
      observationsToGraph.interpretObservations(singleRequestBody({
        'emails': [
          1,
          2,
        ],
      }));
      observationsToGraph.interpretObservations(singleRequestBody({
        'emails': [
          'test@optic.com',
        ],
      }));

      const graphQueries = new GraphQueries(observationsToGraph.graph);
      const [schemaRootNode] = graphQueries.listNodesByType('schemaRoot');
      expect(schemaRootNode.toJsonSchema()).toMatchSnapshot();
    });
  });
});
