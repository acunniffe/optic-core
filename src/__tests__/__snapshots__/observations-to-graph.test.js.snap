// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservationsToGraph simple API should understand the request body format 1`] = `
Object {
  "properties": Object {
    "emails": Object {
      "items": Object {
        "title": "6c29c0c2",
        "type": "string",
      },
      "title": "d6234db7",
      "type": "array",
    },
  },
  "title": "5a7b1b6a",
  "type": "object",
}
`;

exports[`ObservationsToGraph simple API should understand the request body format with different observations 1`] = `
Object {
  "oneOf": Array [
    Object {
      "properties": Object {
        "emails": Object {
          "items": Object {
            "oneOf": Array [
              Object {
                "title": "6c29c0c2",
                "type": "string",
              },
              Object {
                "title": "d47d4fb4",
                "type": "number",
              },
            ],
          },
          "title": "d6234db7",
          "type": "array",
        },
      },
      "title": "5a7b1b6a",
      "type": "object",
    },
    Object {
      "items": Object {},
      "title": "6c1c76c6",
      "type": "array",
    },
    Object {
      "title": "c3b347f8",
      "type": "string",
    },
  ],
}
`;
